controllers/machine_controller.go

```diff
@@ -647,38 +647,70 @@ func (r *MachineReconciler) nodeToMachine(o client.Object) []reconcile.Request {
 	}
 
 	// Match by nodeName and status.nodeRef.name.
-	filters := []client.ListOption{
+	allFilters := []client.ListOption{
 		client.MatchingFields{clusterv1.MachineNodeNameIndex: node.Name},
 	}
+	annotationfilters := []client.ListOption{}
 
 	// Match by clusterName when the node has the annotation.
 	if clusterName, ok := node.GetAnnotations()[clusterv1.ClusterNameAnnotation]; ok {
-		filters = append(filters,
-			client.MatchingLabels{
-				clusterv1.ClusterLabelName: clusterName,
-			},
-		)
+		clusterNameFilter := client.MatchingLabels{
+			clusterv1.ClusterLabelName: clusterName,
+		}
+		annotationfilters = append(annotationfilters, clusterNameFilter)
+		allFilters = append(allFilters, clusterNameFilter)
 	}
 
 	// Match by namespace when the node has the annotation.
 	if namespace, ok := node.GetAnnotations()[clusterv1.ClusterNamespaceAnnotation]; ok {
-		filters = append(filters, client.InNamespace(namespace))
+		namespaceFilter := client.InNamespace(namespace)
+		annotationfilters = append(annotationfilters, namespaceFilter)
+		allFilters = append(allFilters, namespaceFilter)
 	}
 
 	machineList := &clusterv1.MachineList{}
 	if err := r.Client.List(
 		context.TODO(),
 		machineList,
-		filters...); err != nil {
+		allFilters...); err != nil {
 		return nil
 	}
 
 	// There should be exactly 1 Machine for the node.
-	if len(machineList.Items) != 1 {
+	if len(machineList.Items) == 1 {
+		return []reconcile.Request{{NamespacedName: util.ObjectKey(&machineList.Items[0])}}
+	}
+
+	machineList = &clusterv1.MachineList{}
```

Comment:
Add a comment above here that this is effectively a fallback scenario in case the Machine doesn't have a node name index yet.